Add options to javascript and stylesheet helpers, including support for Subresource Integrity (SRI)

accuser opened the pull request:
This PR adds an optional HTML5 `async` attribute to the `#javascript` asset helper. If specified, the resulting HTML output will include the `async` attribute.

davydovanton commented:
LGTM :+1:

joneslee85 commented on test/helpers_test.rb:
```diff
@@ -19,6 +19,11 @@
 
+    it 'renders <script> tag with \'async\' option' do
+      actual = DefaultView.new.javascript('feature-a', async: true)
+      actual.must_equal %(<script src="/assets/feature-a.js" type="text/javascript" async="async"></script>)
```
Can you please test the case when async is false?

accuser replied:
Yep, I can add that, just to prove the default.

> On 14 Feb 2016, at 21:56, Trung Lê <notifications@github.com> wrote:
> 
> In test/helpers_test.rb <https://github.com/hanami/assets/pull/46#discussion_r52849409>:
> 
> > @@ -19,6 +19,11 @@
> >        actual.must_equal %(<script src="/assets/feature-a.js" type="text/javascript"></script>)
> >      end
> >  
> > +    it 'renders <script> tag with \'async\' option' do
> > +      actual = DefaultView.new.javascript('feature-a', async: true)
> > +      actual.must_equal %(<script src="/assets/feature-a.js" type="text/javascript" async="async"></script>)
> Can you please test the case when async is false?
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/hanami/assets/pull/46/files#r52849409>.
>

joneslee85 commented on lib/hanami/assets/helpers.rb:
```diff
@@ -125,9 +131,11 @@ def self.included(base)
       #
       #   # <script src="https://assets.bookshelf.org/assets/application-28a6b886de2372ee3922fcaf3f78f2d8.js" type="text/javascript"></script>
-      def javascript(*sources)
+      def javascript(*sources, **options)
```
What do you think about a default hash `async: true` instead?

And why do you implement the option here at this high level than in the lower level of `script`?

accuser replied:
I don’t think it should be `true` by default, since the `async` option needs to be set, i.e. it is a positive decision.

The option needs to be implemented here otherwise it won’t be available to the helper. The underlying `#script` in the HTML helper could be extended as well, but that is obviously outside the scope of this patch (being that it is in a different repository). Ultimately, the option should be available to both, and I can add it to `hanami-helpers` too.
  
> On 14 Feb 2016, at 22:01, Trung Lê <notifications@github.com> wrote:
> 
> In lib/hanami/assets/helpers.rb <https://github.com/hanami/assets/pull/46#discussion_r52849491>:
> 
> > @@ -125,9 +131,11 @@ def self.included(base)
> >        #   <%= javascript 'application' %>
> >        #
> >        #   # <script src="https://assets.bookshelf.org/assets/application-28a6b886de2372ee3922fcaf3f78f2d8.js" type="text/javascript"></script>
> > -      def javascript(*sources)
> > +      def javascript(*sources, **options)
> What do you think about a default hash async: true instead?
> 
> And why do you implement the option here at this high level than in the lower level of script?
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/hanami/assets/pull/46/files#r52849491>.
>

joneslee85 replied:
Sorry that I did not make myself clear, what I meant is that instead of doing generic wilcard `**options*`, what do you think about explicitly one hash option `async: <what ever default value it is>`?

joneslee85 replied:
Sorry that I did not make myself clear, what I meant is that instead of doing generic wilcard `**options*`, what do you think about explicitly one hash option `async: <what ever default value it is>`?

jodosha commented:
@accuser Thank you for this implementation, but this is too narrowed. `<script>` [accepts at least 8 attributes](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script), without considering `id` or `data-*`. With your proposed change, if we want to support all of them, we need to add a kwarg for each of them. That would pollute the signature of `javascript` and put us in a position to always chase browser vendors by adding more and more kwargs.

Don't get me wrong, I think this change should happen, but not with the current implementation. Again, thanks for requesting this enhancement :wink: 

---

What we should aim for is:

  * Target both `javascript` and `stylesheet`.
  * Allow to specify more attributes
  * Be open for extension: if a new HTML attribute is added, ideally we don't need to touch the code.
  * Support `integrity` attribute with special handling: `integrity: true` should generate something like `integrity="sha384-oqVuAfXRKap7fdgcCY5uykM6+R9GqQ8K/uxy9rx7HNQlGYl1kPzQho1wx4JwY8wC"`. This must be different for each generated `<script>`, of course.

Please read more about `integrity` (aka Subresource Integrity) on [Mozilla documentation](https://developer.mozilla.org/en-US/docs/Web/Security/Subresource_Integrity) and [GitHub engineering blog](http://githubengineering.com/subresource-integrity/).

---

@accuser It would be really, really appreciated if you can expand the scope of the current implementation by taking account of what described above. Thank you :+1:

accuser commented:
@jodosha will do. I opted for kwarg second time around, but preferred the options splat for these exact reasons.

accuser commented:
I've added support for subresource integrity (SRI) on both javascript and stylesheets, as well as open support for HTML attributes. This PR makes a significant change to the asset manifest, but as this would normally be generated for each deploy, this should not be an issue.

I have only added support for a single SRI algorithm (being SHA256), but I have implemented the configuration to allow for addition of multiple algorithms.

accuser commented on lib/hanami/assets/bundler.rb:
```diff
@@ -116,8 +128,11 @@ def generate_manifest
-        @manifest[_convert_to_url(::File.expand_path(asset))] = _convert_to_url(::File.expand_path(target))
+      def store_manifest(asset, target, sri_value)
+        @manifest[_convert_to_url(::File.expand_path(asset))] = {
+          target: _convert_to_url(::File.expand_path(target)),
```
The DigestManifest is a map of assets to their respective digests:

```json
{"/assets/application.js":"/assets/application-123abc.js"}
```

I have extended this to store the SRI attribute value:

```json
{"/assets/application.js":{"target":"/assets/application-123abc.js","sri":"sha256-def456"}}
```

I think that it would be a good idea to drop the `target` for the actual MD5 checksum, generating the asset digest on the fly:

```json
{"/assets/application.js":{"md5":"123abc","sri":"sha256-def456"}}
```

This means that the DigestManifest has a consistent purpose of storing digest/hashes/checksums.

Thoughts?

jodosha replied:
@accuser I've seen that the configuration can handle multiple hashing algorithms, but the manifest reports only one `"sri"` value. Shall we change and make the value an array?

```json
{"/assets/application.js":{"target":"/assets/application-123abc.js","sri":["sha256-def456"]}}
```

jodosha replied:
@accuser 
> I think that it would be a good idea to drop the target for the actual MD5 checksum, generating the asset digest on the fly:

That isn't practical, because at the runtime we should separate the base name from the extension to inject the checksum value. But the problem is to get the right separation (eg `/assets/jquery.min.js`). We already handle that, while compiling. We should duplicate that logic here to guarantee correctness..

Oh, and don't forget that this would be more time consuming, than just fetch `"target"` from the JSON.

My opinion is to leave it as it is :wink:

accuser commented on lib/hanami/assets/configuration.rb:
```diff
@@ -57,6 +57,10 @@ class Configuration
 
+      # @since 0.3.0-add-options-to-javascript-helper
+      # @api private
+      DEFAULT_SRI_ALGORITHM    = 'sha256'.freeze
```
I've chosen SHA256 as the default algorithm, but I have seen examples of SHA384 and SHA512 being used with existing CDNs. I've implemented `#sri_algorithm` to return an array so that more specified through the configuration in future (not yet implemented).

accuser commented on lib/hanami/assets/configuration.rb:
```diff
@@ -458,7 +497,7 @@ def reset!
       # @since 0.1.0
       def load!
-        if digest && manifest_path.exist?
+        if (digest || sri) && manifest_path.exist?
```
A manifest is used if either `digest` or `sri` are `true`.

accuser commented on lib/hanami/assets/helpers.rb:
```diff
@@ -125,9 +153,23 @@ def self.included(base)
+          end
+
+          html.script(
+            src: tag_options.delete(:src),
```
I've done this to maintain the order of attributes, and not break the existing tests.

accuser commented on lib/hanami/assets/helpers.rb:
```diff
@@ -186,9 +240,25 @@ def javascript(*sources)
+          end
+
+          html.link(
+            href: tag_options.delete(:href),
```
I've done this to maintain the order of attributes, and not break the existing tests.

jodosha replied:
@accuser I know it's annoying to change existing tests, but this creates an overhead of instantiate a Hash and we can avoid it.

Do you think that giving a specific order to the Hash saves us from changing the tests? If not, please change the tests. It's a boring task, but it will pay in terms of performance. Thanks for understanding. :smile:

accuser replied:
@jodosha okay

accuser replied:
@jodosha if I am going to be changing the tests, I have a little more scope with the implementation. There are a couple of things that I would like to change, for example, `Hanami::Assets::Bundler#checksum` actually copies the asset file to the digest, which isn't implied by the method. I think that I can refactor `#run` with a couple of more steps which will allow a cleaner implementation of the digest and SRI features. Are you happy for me to go ahead an do this?

jodosha replied:
@accuser Please proceed with the path that you think it's right :+1:

accuser replied:
@jodosha very zen. :wink:

accuser commented on test/fixtures/deploy/assets-yui.json:
```diff
@@ -1 +1 @@
-{"/assets/bootstrap.js":"/assets/bootstrap-1772ab74c1f593547907dc74235f33da.js","/assets/bootstrap.css":"/assets/bootstrap-398412f60b889d9618cd624fdc913eb3.css","/assets/bootstrap-theme.css":"/assets/bootstrap-theme-0a435682ec3e7fe26afdf79c9558a6dd.css","/assets/admin/ui-icons_ffffff_256x240.png":"/assets/admin/ui-icons_ffffff_256x240-a19fe1b2d726e920c12e43a65df9039a.png","/assets/admin/ui-icons_ ...
\ No newline at end of file
+{"/assets/admin/jquery-ui.css":{"target":"/assets/admin/jquery-ui-f7ce8506ccf0bc343c9d7237763c6bfd.css", "sri":"sha256-PvAi7pcBgZlvUlQTYlUIuUWwNCZ3mG5d3X/D6EIFQD4="}, "/assets/admin/jquery-ui.js":{"target":"/assets/admin/jquery-ui-aca7dcb191c5101116071b5044c9e8e6.js", "sri":"sha256-hfzjJw/u5G+RN4nuJ4GS3HfzzrI7cj3ldF1oGI0chnk="}, "/assets/admin/jquery-ui.structure.css":{"target":"/assets/admin/jqu ...
```
I regenerated this file to include the SRI attributes, but unlike the others, it resulted in the content being in a different order.

jodosha commented on lib/hanami/assets/helpers.rb:
```diff
@@ -682,6 +752,15 @@ def _typed_asset_path(source, ext)
+
+      def _sri_value(source, ext)
+        source = "#{ source }#{ ext }" unless source.match(/#{ Regexp.escape(ext) }\z/)
+        self.class.assets_configuration.sri_value(source) unless _absolute_url?(source)
```
@accuser I assume that this guard against absolute URLs is to not try to generate SRI values when we use jQuery CDN. Eg. `https://code.jquery.com/jquery-2.2.0.min.js`

Is that correct?

accuser replied:
Yes it is, although in that case an explicit SRI value can be included via the helper:

```ruby
<%= javascript 'https://code.jquery.com/jquery-2.2.0.min.js', sri: 'sha512-hash-provided-by-jquery-developers' %>
```

jodosha replied:
@accuser `:sri`? Shouldn't it be `:integrity`?

accuser replied:
@jodosha I opted for `sri` in the configuration as this was the overarching feature. Whilst `integrity` is a noun, on its own it doesn't convey the appropriate meaning. However, we can call it whatever we want. :smile:

jodosha replied:
@accuser I'd stick with HTML attributes:

Return the value from the manifest:

```erb
<%= javascript 'application', integrity: true %>
```

Disable for this tag:

```erb
<%= javascript 'application', integrity: false %>
```

Force a different value:

```erb
<%= javascript 'application', integrity: 'sha512-abcdef' %>
```

Does it makes sense?

accuser replied:
@jodosha that's how my implementation works! I the `sri:` in the above example is a copy-n-pasta mistake: I've edited the comment.

accuser replied:
@jodosha when I looked at your comment, I thought you meant the guard method `#_sir?`.

jodosha replied:
@accuser In my understanding `_sri?` is at the configuration level, while that `:integrity` attribute is fine grained control. Does it works like that?

accuser replied:
@jodosha `:sri` is the configuration option, `:_sri?` is a utility method that determines if SRI is enabled, and `:integrity` is the option passed to the `javascript` method that specifies an SRI digest hash for the script (of stylesheet). So, yes, you're right. :smile:

jodosha replied:
@accuser Great to hear. 

> `:sri` is the configuration option

Which configuration option? :smiley_cat:

accuser replied:
@jodosha I added an option in `Hanami::Assets::Configuration` for SRI, following the pattern of the `:digest` option. At the moment it is a simple on or off, but I intend for it to take an optional array of SRI algorithms:

```ruby
assets do
  sri :sha256, :sha512
end
```

The intention would be to generate an SRI digest hash with both algorithms.

jodosha replied:
> The intention would be to generate an SRI digest hash with both algorithms.

Why more than one algorithm at once? :smile:

jodosha replied:
@accuser Is `subresource_integrity` be more explicative? Don't worry about the verbosity, we have code generators for that. :wink:

accuser replied:
@jodosha I looked at how SRI was being used already, and a number of CDN hosts are using multiple SRI digest hash algorithms. Case in point, BootstrapCDN:

```html
<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.6/css/bootstrap.min.css" rel="stylesheet" integrity="sha256-7s5uDGW3AHqw6xtJmNNtr+OBRJUlgkNJEo78P4b0yRw= sha512-nNo+yCHEyn0smMxSswnf/OnX6/KwJuZTlNZBjauKhTK0c+zT+q5JOCx0UFhXQ6rJR9jg6Es8gPuD2uZcYDLqSw==" crossorigin="anonymous”>
```

I would prefer `subresource_integrity` as the option, so will go with that.

accuser replied:
> we have code generators for that

But only once I've coded the template... :laughing:

jodosha commented on lib/hanami/assets/helpers.rb:
```diff
@@ -186,9 +240,25 @@ def javascript(*sources)
+      def stylesheet(*sources, **options)
         _safe_tags(*sources) do |source|
-          html.link(href: _typed_asset_path(source, STYLESHEET_EXT), type: STYLESHEET_MIME_TYPE, rel: STYLESHEET_REL).to_s
+          tag_options = options.merge({
```
@accuser Shouldn't this be the opposite? With this implementation, `href`, `type` and `rel` attributes can't be overwritten by developers.

accuser replied:
@jodosha I tried to remain as close as possible to the original implementation, and this was true then. If a developer wanted to be overwrite the `href`, `type`, and `rel` attributes, they could simply use the HTML helper. However, it could be possible to extend this to allow that sort of fine-grained control, although there would be the need to guard against specific misuse. For example, it wouldn't make sense for `href` to be specified when multiple sources were provided, for example.

jodosha replied:
@accuser Can we avoid to create new Hash instances?

```ruby
options[:href] ||= _typed_asset_path(source, STYLESHEET_EXT)
options[:type] ||= STYLESHEET_MIME_TYPE
options[:rel] ||= STYLESHEET_REL
```

accuser replied:
@jodosha originally I just mutated the hash, but it felt messy. I'll look at an alternative.

jodosha replied:
@accuser If you don't want to mutate the original set of options (:+1:), just `dup` it.

jodosha replied:
@accuser 

> For example, it wouldn't make sense for `href` to be specified when multiple sources were provided, for example.

That is the reason why these options weren't included in the initial implementation. It's handy to have a signature that allows to include multiples assets at once:

```erb
<%= javascript 'jquery', 'application' %>
```

...but a bit awkward to specify single options:

```erb
<%= javascript 'jquery', 'application', id: 'myscript' %>
```

This would generate invalid markup because both the scripts will receive the same DOM ID.

---

We're at a forking path:

  1. Allow developers to use `javascript` to specify one or more scripts, by leaving to them the responsibility of not making mistakes.

  2. Be strict and make `javascript` to generate only one `<script>` tag at once.

I'm for 1 and document this behavior in guides.

---

In any case we should allow them to force `href`, `type` and `rel` attributes. We can never tell which needs a developer can have, we should work to second their needs in this case.

It would be odd to write the following code and get something unexpected in return:

```erb
<%= javascript 'entry-template', type: 'text/x-handlebars-template' %>
```

accuser replied:
@jodosha I agree that 1 is the right path, and developer should be responsible for their actions. :smile: 

I've already made changes (uncommitted) to allow the full scope of attributes to be overloaded. What I could do is add a simple sanity check where multiple sources are provided along with asset specific attributes, such as in your example:

```erb
<%= javascript 'jquery', 'application', id: 'my script' %>
``` 

The above could log a warning.

jodosha replied:
> The above could log a warning.

There are cases where an attribute can be safely specified for multiple tags: think of `class`. That would make `:id` an exception in the way we threat attributes. I'm for leaving things as they are for the purpose of simplify the internals and so have less/faster code.

accuser replied:
> I'm for leaving things as they are for the purpose of simplify the internals and so have less/faster code.

Sweet. Developers be warned. :wink:

I'll make sure that I explicitly document this behaviour in the helpers to mitigate the potential issues.

jodosha commented on lib/hanami/assets/configuration.rb:
```diff
@@ -378,6 +396,25 @@ def asset_url(source)
+        if sri
+          result = prefix.join(source)
+          result = digest_manifest.resolve(result)
+          result.fetch('sri').to_s
```
@accuser because we can have more than one SRI values, this should do `result.fetch('sri').join(" ")`.

jodosha replied:
@accuser Similarly to https://github.com/hanami/assets/pull/46/files#r55831013, we should probably add a new method:

```ruby
result = digest_manifest.sri(result)
```

```ruby
def sri(asset)
  resolve(asset).fetch('sri')
end
```

WDYT?

jodosha commented on lib/hanami/assets/configuration.rb:
```diff
@@ -524,7 +567,12 @@ def load!
-        result = digest_manifest.resolve(result) if digest
+
+        if digest
+          result = digest_manifest.resolve(result)
```
@accuser We're leaking internal informations of `DigestManitfest`. We should think to expose a new method `#target`. So we can do:

```ruby
result = digest_manifest.target(result) if digest
```

It's implementation should be

```ruby
def target(path)
  resolve(result).fetch('target')
end
```

jodosha commented:
I'm closing this PR as it can't be merged **and** because, I want to take this code and cleanup. @accuser Thank you very much. Your work is a solid base for this feature. 💚
